src/lib/device-types/sync/index.ts

```diff
@@ -0,0 +1,104 @@
+import * as _ from 'lodash';
+import Bluebird = require('bluebird');
+import { captureException } from '../../../platform/errors';
+import { sbvrUtils, permissions } from '@balena/pinejs';
+import { getAllDeviceTypes } from '..';
+import { DeviceType } from '../build-info-facade';
+
+const DEVICE_TYPES_CACHE_EXPIRATION = 5 * 1000; // 5 mins
+const RETRY_WAIT_TIME = 10 * 1000;
+
+type DeviceTypeResourceKeys = 'slug' | 'name' | 'is_private';
+
+// This map will hold information on which device type fields imported from the device type registry will be synced to which db fields.
+export const syncSettings: {
+	map: {
+		[k in DeviceTypeResourceKeys]?: {
+			name: keyof DeviceType;
+			default?: any;
+		};
+	};
+} = {
+	map: { slug: { name: 'slug' } },
+};
+
+export function setSyncMap(map: typeof syncSettings['map']) {
+	syncSettings.map = _.merge(map, syncSettings.map);
+}
+
+const mapModel = (deviceType: DeviceType, map: typeof syncSettings['map']) => {
+	return Object.keys(map).reduce(
+		(
+			mappedModel: { [k in DeviceTypeResourceKeys]?: any },
+			key: Partial<DeviceTypeResourceKeys>,
+		) => {
+			mappedModel[key] = deviceType[map[key]!.name] ?? map[key]!.default;
+			return mappedModel;
+		},
+		{},
+	);
+};
+
+const syncDeviceTypes = async () => {
+	const deviceTypes = await getAllDeviceTypes();
+	const mappedDeviceTypes = deviceTypes.map((deviceType) =>
+		mapModel(deviceType, syncSettings.map),
+	);
+
+	return sbvrUtils.db.transaction(async (tx) => {
+		const rootApi = sbvrUtils.api.resin.clone({
+			passthrough: { req: permissions.root, tx },
+		});
+
+		const existingDeviceTypes = (await rootApi.get({
+			resource: 'device_type',
+			options: { $select: ['slug'] },
+		})) as Array<{ slug: string }>;
+
+		const existingSlugs = new Set(
+			existingDeviceTypes.map((deviceType) => deviceType.slug),
+		);
+
+		return Bluebird.map(
+			mappedDeviceTypes,
+			async (deviceType) => {
+				try {
+					if (existingSlugs.has(deviceType.slug)) {
+						return await rootApi.patch({
+							resource: 'device_type',
+							body: deviceType,
+							options: {
+								$filter: { slug: deviceType.slug },
+								returnResource: false,
+							},
+						});
+					}
+
+					return await rootApi.post({
+						resource: 'device_type',
+						body: deviceType,
+						options: { returnResource: false },
+					});
+				} catch (err) {
+					captureException(
+						err,
+						`Failed to synchronize device type: ${deviceType.slug}, skipping...`,
+					);
```

Comment:
If a request within a transaction fails then you can't really ignore the error because it'll almost certainly roll back the transaction and cause any other db calls to fail

Reply:
Yeah makes sense, I just didn't want to fail all DT sync because of one having issues in the device-type.json for example, but I guess it's better to fail fast.